docs: add how-to for sending logs to external rsyslog via OTel Collector Integrator (#313)
canonical/observability-stack: docs/how-to/integrate/send-telemetry-to-external-backends.md

Reviewer: In its current form, this is a valuable _tutorial_.
How about we move this file into tutorials, and then extract a how-to from the tutorial?

---
myst:
  html_meta:
    description: "Send logs to an external rsyslog server over TLS using the OpenTelemetry Collector Integrator charm with Juju secrets for secure credential management."
---

# How to send logs to non-charmed backends

The [OpenTelemetry Collector Integrator](https://github.com/canonical/opentelemetry-collector-integrator-operator) charm enables you to export telemetry from charmed OpenTelemetry Collector to non-charmed backends. It acts as a configuration add-on that injects arbitrary exporter configurations, backed by Juju secrets.

> This obviates the need for [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) or similar other solutions.

Reviewer 2: ```suggestion
This removes the need for [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) or similar other solutions.
```
"removes" is more plain English

> 

Reviewer 2: Add another line to make it clear what's in the doc, such as "This guide demonstrates how to send logs to an external rsyslog server over TLS using the OpenTelemetry Collector Integrator."

Just to make it more clear what the user will actually be guided to do here

```

## Architecture overview

```{mermaid}
graph LR

subgraph Juju model
  app --- opentelemetry-collector
  opentelemetry-collector-integrator ---|"external-config"| opentelemetry-collector
end

opentelemetry-collector -.-|"syslog/rfc5424"| non-charmed[Non-charmed e.g. rsyslog]

juju-config@{ shape: braces, label: "Admin config<br> for rsyslog" } -.->|juju config| opentelemetry-collector-integrator
juju-action@{ shape: braces, label: "Optional secrets<br> for rsyslog" } -.->|juju run| opentelemetry-collector-integrator
```

The admin needs to provide a valid opentelemetry exporter config (example: [syslog](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/syslogexporter)). Once the `external-config` integration between opentelemetry-collector and the integrator charms is established,
opentelemetry-collector will merge the injected configuration and begin forwarding logs.

The integrator charm may be related [cross-model](https://documentation.ubuntu.com/juju/3.6/howto/manage-relations/#add-a-cross-model-relation).

> TODO: add a link to the tutorial.

Reviewer 2: note this line needs to change

> 

Reviewer 2: Any important prerequisites? (presumably they already need to have otelcol setup?) If yes, you can add another section here ("Prerequisites", "Before you begin", etc....)

Asking because it wasn't clear to me at what stage of their deployment setup would a user be following this guide. If you think it's already clear to your users, disregard this

Reviewer 2: Thinking about it more, maybe your prereqs could be like:

"
- A working COS deployment with:

   - The OpenTelemetry Collector deployed
   - At least one machine charm sending logs
  
- (cont'd with other specs if necessary)
"

> ## Create relevant secrets

Reviewer 2: The deploy section from the tutorial could be refactored and added before this section, if it's not included in the prereqs. (also is there a `juju relate` missing?)

---
myst:
  html_meta:
    description: "Send logs to an external rsyslog server over TLS using the OpenTelemetry Collector Integrator charm with Juju secrets for secure credential management."
---

# How to send logs to non-charmed backends

> The [OpenTelemetry Collector Integrator](https://github.com/canonical/opentelemetry-collector-integrator-operator) charm enables you to export telemetry from charmed OpenTelemetry Collector to non-charmed backends. It acts as a configuration add-on that injects arbitrary exporter configurations, backed by Juju secrets.

Reviewer 3: ```suggestion
The [OpenTelemetry Collector Integrator](https://github.com/canonical/opentelemetry-collector-integrator-operator) charm enables exporting telemetry from the charmed OpenTelemetry Collector to non-charmed backends. It acts as a configuration add-on that injects arbitrary exporter configurations, backed by Juju secrets.
```

Is this true?
> a configuration add-on that injects arbitrary exporter configurations

technically, we allow them to configure more than just exporters ...

This removes the need for [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) or similar other solutions.

> This guide walks through sending logs to an external rsyslog server over TLS, demonstrating how the Integrator handles sensitive material like certificates through its secret templating system.

Reviewer 2: Presumably an external rsyslog server is an example of a "non-charmed backend" ?

Asking because there seems to be a mismatch between the title and what this says.

If it's meant to be an example, you could add another sentence here too, so the section is something like

```suggestion
This guide walks through sending logs to an external rsyslog server over TLS, demonstrating how the Integrator handles sensitive material like certificates through its secret templating system. The general approach here can be used with other backends.
```

Author: Yes, you are right! Thanks @YanisaHS

```{important}
Available components (exporters, receivers, processors) depend on the OpenTelemetry Collector build compiled by Canonical. Check the [manifest-additions.yaml](https://github.com/canonical/opentelemetry-collector-rock) in the opentelemetry-collector-rock repository for your version to confirm what's included.
```

## Architecture overview

```{mermaid}
graph LR

subgraph Juju model
  app --- opentelemetry-collector
  opentelemetry-collector-integrator ---|"external-config"| opentelemetry-collector
end

opentelemetry-collector -.-|"syslog/rfc5424"| non-charmed[Non-charmed e.g. rsyslog]

juju-config@{ shape: braces, label: "Admin config<br> for rsyslog" } -.->|juju config| opentelemetry-collector-integrator

> juju-action@{ shape: braces, label: "Optional secrets<br> for rsyslog" } -.->|juju run| opentelemetry-collector-integrator

Reviewer 3: Thoughts on being more explicit about the charm action and config names?
```suggestion
juju-config@{ shape: braces, label: "Admin config<br> for rsyslog" } -.->|juju config\nconfig_yaml| opentelemetry-collector-integrator
juju-action@{ shape: braces, label: "Optional secrets<br> for rsyslog" } -.->|juju run\ncreate-secret| opentelemetry-collector-integrator
```

```

The Integrator injects a syslog exporter and a transform processor into the Collector's logs pipeline. Once the `external-config` integration is established, the Collector automatically merges the injected configuration and begins forwarding logs.

## Prerequisites

- A Juju controller and model on a machine cloud (for example, LXD)
- The `ubuntu` charm deployed in the model

> - The `opentelemetry-collector` charm deployed as a subordinate of `ubuntu`, named `otelcol` for short.

Reviewer 2: I'm not sure what you're going for here - but this sounds like a prerequisite is that the user should have `opentelemetry-collector` re-named as `otelcol` in their deployment (i.e., if they didn't already name it `otelcol`, they should change it before starting the guide).

If that's what you want, then it's okay, but you could also change it to "The `opentelemetry-collector` charm deployed as a subordinate of `ubuntu`. This guide names the charm `otelcol` for short." - if you meant that this guide will refer to it as `otelcol`.

- An external rsyslog server reachable from the Juju machines, configured to accept TLS connections on port `6514`
- TLS certificate material for mutual TLS: CA certificate, client certificate, and client key
- `juju` CLI version 3.6 or later

```{note}
This guide uses the machine (VM) variant of the charms. For Kubernetes, replace `opentelemetry-collector` with `opentelemetry-collector-k8s`.
```

## Step 1: Deploy the OpenTelemetry Collector Integrator

Deploy the Integrator charm into the same model where `otelcol` is running:

```bash
juju deploy opentelemetry-collector-integrator otelcol-integrator --channel latest/edge
```

> Wait for it to settle:

Reviewer 2: IMO this isn't clear what it means. What does it mean for it to be "settled"? (rather than waiting for X thing to have X output)

If your users understand this, you can ignore it though. But pointing it out still

Author: Good point!

```bash
juju status --watch 2s
```

> The Integrator enters `blocked` status until you provide a valid configuration with at least one pipeline enabled. This is expected.

Reviewer 3: ```suggestion
The Integrator enters `blocked` status until you provide a valid configuration with at least one pipeline enabled.
```

- `network`: must be `tcp` for TLS
- `protocol`: `rfc5424` (modern syslog standard)
- `tls`: references to the Juju secret using the `?render=file` directive — the Collector writes each certificate to disk with restricted permissions and substitutes the file path

The secret URI format is: `<secret-id>/<key>?render=<type>`, where `<secret-id>` is the full value returned by the `create-secret` action, `<key>` is one of the keys you stored in the secret, and `<type>` is `file` (for certificates written to disk) or `inline` (for values substituted directly in the config text).
```

## Step 4: Apply the configuration

Configure the Integrator with the exporter YAML and enable the logs pipeline:

```bash
juju config otelcol-integrator \
    config_yaml=@syslog-tls-exporter.yaml \
    logs_pipeline=true
```

Verify the Integrator reaches `active` status:

```bash
juju status otelcol-integrator
```

You should see:

```
App                 Version  Status   Scale  Charm                               Channel        Rev  Exposed  Message
otelcol-integrator           active       1  opentelemetry-collector-integrator  latest/edge      3  no       Pipelines: logs configured
```

> If the charm shows `blocked`, check `juju debug-log --include otelcol-integrator` for validation errors in your YAML or pipeline configuration.

Reviewer 3: Replay for more context, or is this too much handholding?
```suggestion
If the charm shows `blocked`, check `juju debug-log --include otelcol-integrator --replay` for validation errors in your YAML or pipeline configuration.
```

## Step 5: Integrate the Collector with the Integrator

Establish the `external-config` integration:

```bash
juju integrate otelcol:external-config otelcol-integrator:external-config
```

Once the integration is established, the Integrator shares the configuration and grants access to the Juju secret. The Collector automatically:

1. Retrieves the secret content
2. Writes certificate files to disk (for `?render=file` references)
3. Merges the exporter and processor into its logs pipeline
4. Reloads its configuration

```{note}

> The `otelcol` application  may remain in `blocked` status because the `external-config` integration does not satisfy the charm's primary outbound integration requirements (such as `send-loki-logs` or `send-otlp`). This does not prevent the external exporter from functioning — logs are still forwarded to rsyslog.

Reviewer 3: Can you create an issue for this to update the mandatory-relation pair logic to not block the charm (K8s and VM) if a user-defined config is enabled, because we should trust that the user has configured this correctly? I feel like this introduces bad UX. Isn't this solved by Simme's signal studio?

Author: Good point!

https://github.com/canonical/opentelemetry-collector-operator/issues/273

---
myst:
  html_meta:
    description: "Send logs to an external rsyslog server over TLS using the OpenTelemetry Collector Integrator charm with Juju secrets for secure credential management."
---

# How to send logs to an external rsyslog server

> The [OpenTelemetry Collector Integrator](https://github.com/canonical/opentelemetry-collector-integrator-operator) charm enables you to export telemetry from Juju-managed OpenTelemetry Collector instances to backends that do not have native Juju charms. It acts as a secure configuration bridge that injects exporter configurations and manages credentials as Juju secrets.

Reviewer: ```suggestion
The [OpenTelemetry Collector Integrator](https://github.com/canonical/opentelemetry-collector-integrator-operator) charm enables you to export telemetry from charmed OpenTelemetry Collector to non-charmed backends. It acts as a configuration add-on that injects arbitrary exporter configurations, backed by Juju secrets.
```

> For metrics specifically, this provides a push-based alternative to [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) — instead of configuring hierarchical scrape chains, the Collector forwards metrics directly to a remote backend via any supported exporter (such as `prometheusremotewrite` or `otlp`).

Reviewer: ```suggestion
For metrics specifically, this provides a push-based alternative to [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) — instead of configuring hierarchical scrape chains, the Collector forwards metrics directly to a remote backend via any supported exporter (such as `rsyslog`).
```

Also add link somewhere to the exporters (contrib?) repo?

This guide walks through sending logs to an external rsyslog server over TLS, demonstrating how the Integrator handles sensitive material like certificates through its secret templating system.

```{important}

> Available components (exporters, receivers, processors) depend on the OpenTelemetry Collector build compiled by Canonical. Check the [manifest-additions.yaml](https://github.com/canonical/opentelemetry-collector-rock) in the opentelemetry-collector-rock repository for your version to confirm what's included.

Reviewer: For track 3 we can probably like directly to the 0.130 file.

```

## Architecture overview

```{mermaid}
graph BT
    subgraph JujuModel["🔷 Juju Model"]
        Ubuntu["📦 Ubuntu Charm<br/>(Application source)"]
        OtelCol-Integrator["⚙️ OpenTelemetry Collector Integrator"]
        OtelCol["🔄 OpenTelemetry Collector"]
    end

    Rsyslog["📋 External rsyslog server<br/>(TLS on port 6514)"]

Ubuntu ---|"juju-info"| OtelCol
OtelCol-Integrator -->|"external-config"| OtelCol

> OtelCol -->|"syslog/rfc5424<br/>over TLS"| Rsyslog

Reviewer: - Should we use precise charm names here, the same way they are named on charmhub?
- The line to rsyslog should be somehow different to better convey juju relation vs HTTP.

```

> The Integrator injects a syslog exporter and a transform processor into the Collector's logs pipeline. Once the `external-config` integration is established, the Collector automatically merges the injected configuration and begins forwarding logs.

Reviewer: Seems repetitive?

## Prerequisites

- A Juju controller and model on a machine cloud (for example, LXD)
- The `ubuntu` charm deployed in the model

> - The `opentelemetry-collector` charm deployed as a subordinate of `ubuntu`, named `otelcol` for short.

Reviewer: Should we combine this with the mermaid diagram? Seems repetitive otherwise.

- An external rsyslog server reachable from the Juju machines, configured to accept TLS connections on port `6514`
- TLS certificate material for mutual TLS: CA certificate, client certificate, and client key
- `juju` CLI version 3.6 or later

```{note}
This guide uses the machine (VM) variant of the charms. For Kubernetes, replace `opentelemetry-collector` with `opentelemetry-collector-k8s`.
```

## Step 1: Deploy the OpenTelemetry Collector Integrator

Deploy the Integrator charm into the same model where `otelcol` is running:

```bash

> juju deploy opentelemetry-collector-integrator otelcol-integrator --channel latest/edge

Reviewer: Can we do stable?
